Add http client and insecure tls tests for registry client

runyontr opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md

Needed a big refactor of https://github.com/helm/helm/pull/9564 after OCI moved away from experimental.  Want to give @chenwng credit for most of the implementation details.

**What this PR does / why we need it**:

Adds features to the registry client to allow:

* HTTP OCI registries
* custom CAs
* key/cert authentication

Will enable features being added to the CLI when ready

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

souleb committed:
Enable custom certificates option for OCI

If implemented, users will be able to use custom certificates and CA to
while interacting with OCI registries.

Signed-off-by: Soule BA <bah.soule@gmail.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

souleb committed:
Enable custom certificates option for OCI

If implemented, users will be able to use custom certificates and CA to
while interacting with OCI registries.

Signed-off-by: Soule BA <bah.soule@gmail.com>

dependabot[bot] committed:
chore(deps): bump github.com/spf13/cobra from 1.5.0 to 1.6.1

Bumps [github.com/spf13/cobra](https://github.com/spf13/cobra) from 1.5.0 to 1.6.1.
- [Release notes](https://github.com/spf13/cobra/releases)
- [Commits](https://github.com/spf13/cobra/compare/v1.5.0...v1.6.1)

---
updated-dependencies:
- dependency-name: github.com/spf13/cobra
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Wonong committed:
Add a flag var to check git is installed or not

Signed-off-by: Wonyeong Choi <ciwnyg0815@gmail.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

phil9909 committed:
improve error message on plugin install

Signed-off-by: Philipp Stehle <philipp.stehle@sap.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Wonong committed:
Update logic of non-git situation just to print warning logs

Signed-off-by: Wonyeong Choi <ciwnyg0815@gmail.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

sabre1041 committed:
Added support for insecure OCI registries

Signed-off-by: Andrew Block <andy.block@gmail.com>

yxxhero committed:
bump sprig version 3.2.3

Signed-off-by: yxxhero <aiopsclub@163.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

sabre1041 committed:
Added insecure option to login subcommand

Signed-off-by: Andrew Block <andy.block@gmail.com>

sabre1041 committed:
Added insecure option to login subcommand

Signed-off-by: Andrew Block <andy.block@gmail.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Commit:
Add insecure, http, and tls settings for oci client

Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Commit:
Fixed tests

Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

sabre1041 commented:
@runyontr thanks for the contribution. There is a collaborative effort ongoing right now with this similar feature that is living on the [oci-transport-refinement](https://github.com/helm/helm/tree/oci-transport-refinement) branch. The latest PR to that branch is https://github.com/helm/helm/pull/11599

sabre1041 committed:
Removed conditional

Signed-off-by: Andrew Block <andy.block@gmail.com>

sabre1041 committed:
Removed conditional

Signed-off-by: Andrew Block <andy.block@gmail.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

runyontr commented:
@sabre1041 Let me know how would be best to collaborate.  Happy close to close mine and get the work into your branch if you want.  I'd like to get the http client work in to unblock some work on fluxcd/source-controller, and don't see that specific capability in your branch.

sabre1041 committed:
Reimplemented change in httpgetter for insecure TLS option

Signed-off-by: Andrew Block <andy.block@gmail.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

sabre1041 committed:
Reimplemented change in httpgetter for insecure TLS option

Signed-off-by: Andrew Block <andy.block@gmail.com>

Commit:
Add HTTP Support for OCI artifacts

Signed-off-by: Tom Runyon <runyontr@gmail.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Commit:
added test for insecure registry

Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Commit:
Removed unused vars and adjusted function parameters

Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Commit:
Revert testing back.  Works locally, not in CI

Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

souleb commented:
`TLS` is already enabled in this branch (see test cases in `client_tls_test.go`.

runyontr commented:
Tracking that TLS was already implemented and working. I believe this adds the following:
* Plain HTTP client
* Plain HTTP Client tests
* Custom CA TLS Client tests
* InsecureSkipVerifyTLS Client Tests

souleb committed:
Provide a helper to set the registryClient in cmd

If enabled the registryClient is set using a helper that accepts the TLS
flags. This keeps the client creation consistent accross the different
commands.

Signed-off-by: Soule BA <bah.soule@gmail.com>

sabre1041 committed:
Merge pull request #11659 from souleb/custom-cert-oci

Provide a helper to set the registryClient in cmd

sabre1041 reviewed (changes requested):
@runyontr Thanks for the contribution as a lot of this content is going to drive the path towards OCI stabilization.

A few notes:

1. Recent integrations by @souleb in #11659 have put the TLS configurations in a good place
2. We are still missing Plain HTTP support so if we can integrate it in with the current state of the branch, that would be great
3. Love seeing the tests associated with this PR, so thank you!
4. CircleCI failures should be resolved soon as the Helm project will be looking to finalize the migration to focus solely on GitHub Actions for CI. PR: https://github.com/helm/helm/pull/11676

Commit:
Save files after merge conflict resolution

Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Commit:
Add HTTP support for OCI artifacts

Signed-off-by: Tom Runyon <runyontr@gmail.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Commit:
Add HTTP Support for OCI artifacts

Signed-off-by: Tom Runyon <runyontr@gmail.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Commit:
add the logic for http back when creating a client
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

sabre1041 reviewed (changes requested):
This is looking good. A few things:

1. Additional tasks are needed wire up to the CLI commands, right?
2. Would you be able to signoff your commits to pass the DCO check?

Commit:
Added flags to CLI commands

Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Commit:
Enable HTTP connection for OCI artifacts

Signed-off-by: Soule BA <bah.soule@gmail.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Commit:
Added support for insecure OCI registries

Signed-off-by: Andrew Block <andy.block@gmail.com>
Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Commit:
Merge remote-tracking branch 'helm/oci-transport-refinement' into oci-insecure

Signed-off-by: Tom Runyon <runyontr@gmail.com>

Commit:
Pass cli configuration around

Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

souleb commented on cmd/helm/pull.go:
```diff
@@ -43,7 +43,7 @@ result in an error, and the chart will not be saved locally.
 
 func newPullCmd(cfg *action.Configuration, out io.Writer) *cobra.Command {
-	client := action.NewPullWithOpts(action.WithConfig(cfg))
+	client := action.NewPullWithOpts(action.WithConfig(cfg), action.WithPullOptWriter(out))
```
no need for `action.WithPullOptWriter`. We don't use it anymore since [11480](https://github.com/helm/helm/pull/11480/files). Everything is redirected to `OS.Stderr`.

souleb commented on cmd/helm/install.go:
```diff
@@ -175,6 +175,7 @@ func addInstallFlags(cmd *cobra.Command, f *pflag.FlagSet, client *action.Instal
 	f.BoolVar(&client.Atomic, "atomic", false, "if set, the installation process deletes the installation on failure. The --wait flag will be set automatically if --atomic is used")
 	f.BoolVar(&client.SkipCRDs, "skip-crds", false, "if set, no CRDs will be installed. By default, CRDs are installed if not already present")
 	f.BoolVar(&client.SubNotes, "render-subchart-notes", false, "if set, render subchart notes along with the parent")
+	f.BoolVar(&client.PlainHTTP, "plain-http", false, "install charts from http registry")
```
I think the `PlainHTTP` flag should be added here https://github.com/helm/helm/blob/866c6af20c1eef2f2db560528072af9cd9ed05ff/cmd/helm/flags.go#L53

It will then be available to the downloading commands.

souleb replied:
```suggestion
	f.BoolVar(&client.PlainHTTP, "plain-http", false, "use plain http for the chart download")
```

souleb commented on cmd/helm/pull.go:
```diff
@@ -87,6 +87,7 @@ func newPullCmd(cfg *action.Configuration, out io.Writer) *cobra.Command {
 	f.BoolVar(&client.VerifyLater, "prov", false, "fetch the provenance file, but don't perform verification")
 	f.StringVar(&client.UntarDir, "untardir", ".", "if untar is specified, this flag specifies the name of the directory into which the chart is expanded")
 	f.StringVarP(&client.DestDir, "destination", "d", ".", "location to write the chart. If this and untardir are specified, untardir is appended to this")
+	f.BoolVar(&client.PlainHTTP, "plain-http", false, "install charts from http registry")
```
```suggestion
	f.BoolVar(&client.PlainHTTP, "plain-http", false, "use plain http for the chart download")
```

souleb commented on cmd/helm/push.go:
```diff
@@ -93,6 +95,7 @@ func newPushCmd(cfg *action.Configuration, out io.Writer) *cobra.Command {
 	f.StringVar(&o.keyFile, "key-file", "", "identify registry client using this SSL key file")
 	f.StringVar(&o.caFile, "ca-file", "", "verify certificates of HTTPS-enabled servers using this CA bundle")
 	f.BoolVar(&o.insecureSkipTLSverify, "insecure-skip-tls-verify", false, "skip tls certificate checks for the chart upload")
+	f.BoolVar(&o.plainHTTP, "plain-http", false, "push charts to registry over http")
```
```suggestion
	f.BoolVar(&o.plainHTTP, "plain-http", false, "use plain http for the chart upload")
```

souleb commented on cmd/helm/show.go:
```diff
@@ -194,6 +194,7 @@ func addShowFlags(subCmd *cobra.Command, client *action.Show) {
 	f := subCmd.Flags()
 
 	f.BoolVar(&client.Devel, "devel", false, "use development versions, too. Equivalent to version '>0.0.0-0'. If --version is set, this is ignored")
+	f.BoolVar(&client.PlainHTTP, "plain-http", false, "use plain http connection instead of https")
```
```suggestion
	f.BoolVar(&client.PlainHTTP, "plain-http", false, "use plain http for the chart download")
```

souleb commented:
Happy new year @runyontr. Made some comments in `cmd`. I also wanted to review `pkg` but it seems not ready yet.

runyontr committed:
Update cmd/helm/push.go

Co-authored-by: souleb <bah.soule@gmail.com>
Signed-off-by: Thomas Runyon <runyontr@gmail.com>

runyontr committed:
Update cmd/helm/pull.go

Co-authored-by: souleb <bah.soule@gmail.com>
Signed-off-by: Thomas Runyon <runyontr@gmail.com>

Commit:
cmd updates based on pr feedback

Signed-off-by: Tom Runyon <tom@defenseunicorns.com>

Commit:
Merge branch 'oci-insecure' of github.com:runyontr/helm into oci-insecure

Signed-off-by: Tom Runyon <tom@defenseunicorns.com>
